src/chkstat.cpp

```diff
@@ -21,1365 +21,1246 @@
+        // make out the leading path component
+        auto sep = path_rest.find_first_of('/', 1);
+        component = path_rest.substr(1, sep - 1);
+        // strip the leading path component
+        path_rest = path_rest.substr(component.length() + 1);
+        // path_rest is empty when we reach the final path element
+        is_final_path_element = path_rest.empty() || path_rest == "/";
 
-        for (i = 1; i < (size_t)argc; i++)
-        {
-            add_checklist(argv[i]);
-            use_checklist = 1;
+        // multiple consecutive slashes: ignore
+        if (!is_final_path_element && component.empty())
             continue;
-        }
-        collect_permfiles();
-    }
-    else if (argc <= 1)
-        usage(1);
-    else
-    {
-        npermfiles = (size_t)argc-1;
-        permfiles = &argv[1];
-    }
 
-    if (have_fscaps == 1 && !check_fscaps_enabled())
-    {
-        fprintf(stderr, "Warning: running kernel does not support fscaps\n");
-    }
+        // never move up from the configured root directory (using the stat result from the previous loop iteration)
+        if (component == ".." && !altroot.empty() && ctx.status.sameObject(root_status))
+            continue;
 
-    if  (do_set == -1)
-        do_set = 0;
+        {
+            // component is an empty string for trailing slashes, open again with different open_flags.
+            auto child = component.empty() ? "." : component.c_str();
+            int tmpfd = openat(pathfd.get(), child, O_PATH | O_NOFOLLOW | O_CLOEXEC | O_NONBLOCK);
+            // TODO: shouldn't there be some error handling here? ENOENT is
+            // probably to be tolerated when packages aren't installed, but
+            // what about e.g. EACCES or other errors?
+            if (tmpfd == -1)
+                return false;
+            pathfd.set(tmpfd);
+        }
 
-    // add fake list entries for all files to check
-    for (i = 0; i < nchecklist; i++)
-        add_permlist(checklist[i], "unknown", "unknown", 0);
+        if (!ctx.status.fstat(pathfd.get()))
+            // TODO: this is also a strange case, should be complained about
+            return false;
 
-    for (i = 0; i < npermfiles; i++)
-    {
-        if ((fp = fopen(permfiles[i], "r")) == 0)
+        // owner of directories must be trusted for setuid/setgid/capabilities
+        // as we have no way to verify file contents
+        //
+        // for euid != 0 it is also ok if the owner is euid
+        if (ctx.status.st_uid && ctx.status.st_uid != m_euid && !is_final_path_element)
         {
-            perror(permfiles[i]);
-            exit(1);
+            ctx.traversed_insecure = true;
         }
-        lcnt = 0;
-        struct perm* last = NULL;
-        int extline;
-        while (readline(fp, line, sizeof(line)))
+        // path is in a world-writable directory, or file is world-writable itself.
+        else if (!ctx.status.isLink() && ctx.status.isWorldWritable() && !is_final_path_element)
```

Comment:
The comment and the code are at odds with each other - `!is_final_path_element` excludes the second option from the comment.

Reply:
Did you think the change of the comment through and arrived at the conclusion that the code is indeed doing the right thing here? (Because I was too lazy for now.)

Reply:
OK, I've put in the work now... so for regular files we have an error after the loop:

https://github.com/openSUSE/permissions/blob/9d7a0af4bdeb22be90521f7be78ff977a06076d3/src/chkstat.cpp#L1028

We don't for directories, which is fine I guess. So just fixing the comment is fine.